### PR TITLE
Fix return value for Manager::parseOpdsDom

### PR DESCRIPTION
**Pull request by BPerlakiH**
Fixes: #1099

### CODE REVIEW

File: src/manager.cpp [hunk 1]
```diff
@@ -143,9 +143,8 @@ bool Manager::parseOpdsDom(const pugi::xml_document& doc, const std::string& url
     m_totalBooks = strtoull(libraryNode.child("totalResults").child_value(), 0, 0);
     m_startIndex = strtoull(libraryNode.child("startIndex").child_value(), 0, 0);
     m_itemsPerPage = strtoull(libraryNode.child("itemsPerPage").child_value(), 0, 0);
-    m_hasSearchResult = true;
```
Comment (mgautierfr) [inline src/manager.cpp:146]: `m_hasSearchResul` is a member of `Manager` and is public, so part of the API.
Even if, indeed, it is not used in any of our project, we cannot simply not set it.

The `totalResults`/`startIndex`/`itemsPerPage` nodes are present when parsing a OPDS which is a search result. If parsing a root opds flux, they are not present.
So we

Comment (mgautierfr) [inline src/manager.cpp:146]: `m_hasSearchResul` is a member of `Manager` and is public, so part of the API.
Even if, indeed, it is not used in any of our project, we cannot simply not set it.

File: src/manager.cpp [hunk 2]
```diff
@@ -143,9 +143,8 @@ bool Manager::parseOpdsDom(const pugi::xml_document& doc, const std::string& url
...
   } catch(...) {
-    m_hasSearchResult = false;
+    return false;
```
Comment (mgautierfr) [inline src/manager.cpp:147]: The `totalResults`/`startIndex`/`itemsPerPage` nodes are present when parsing a OPDS which is a search result. If parsing a root opds stream, they are not present.
So we must not stop the parsing of the opds stream if they are not present.

**Review by veloman-yunkan — COMMENTED** [review]
I am not sure what `kiwix::Manager::parseOpdsDom()` is supposed to return. I am also not sure why that method is declared `protected`. I don't know why the `m_hasSearchResult`, `m_totalBooks`, `m_startIndex`, and `m_itemsPerPage` data members of `kiwix::Manager` are needed and why they are public.

Therefore I think that we should address all of the above comprehensively.

**Review by mgautierfr — CHANGES_REQUESTED** [review]
In itself, this function cannot really fails.
The `doc` is already parsed and all methods call on nodes return [null node](https://pugixml.org/docs/manual.html#node_null) or empty string. (Never throw exception).
And we parse the string with `strtoull` which will always return a value (and set errno in case of error (we should check).

So either we have to deeply modify this function (and `Book::updateFromOpds`) to correctly[1] handle missing node/attribute or the `bool` return is useless and we should change it to `void`.

[1] The next question is "What is correctly`. The current version is tolerant to missing data and setup book with empty attribute in this case. Should we stop parsing of the whole stream if one attribute is missing ? Should we continue but tell (with a bool) that something (without telling what) goes wrong ?

**Comment by mgautierfr** [comment]
> I am not sure what kiwix::Manager::parseOpdsDom() is supposed to return. I am also not sure why that method is declared protected.

It should probably be declared private and return `void` (at least with the current code).

> I don't know why the m_hasSearchResult, m_totalBooks, m_startIndex, and m_itemsPerPage data members of kiwix::Manager are needed and why they are public.

It is not clear even to me :)
As said, opds stream can be also for search result. There is this information in the stream and we must provide it to user one way or the other.

I suppose it was used somehow at a moment but I cannot found where/when.
Maybe I was just (excessively) careful and simply parse it and put it here to have the value but we never used them.

File: src/manager.cpp (hunk 1 above)
Comment (BPerlakiH) [reply src/manager.cpp:146]: Ok, makes sense, I have reverted that part.

**Review by veloman-yunkan — COMMENTED** [review]
This PR doesn't fix anything. The return value of `kiwix::Manager::parseOpdsDom()` is not used in any of kiwix projects. In theory it *could* be used by some code that inherits from `kiwix::Manager` but even then the fix should start from first defining the semantics of the return value of `kiwix::Manager::parseOpdsDom()`. It looks that a proper solution would be to change the return type to `void`, but that would mean a change in the API. A roughly equivalent solution is to state that the return value of `kiwix::Manager::parseOpdsDom()` doesn't matter.

**Comment by kelson42** [comment]
@veloman-yunkan How should we know if parsing went well or not?

**Comment by veloman-yunkan** [comment]
> @veloman-yunkan How should we know if parsing went well or not?

A higher level function `kiwix::Manager::readOpds()` is responsible for that. See [my comment](https://github.com/kiwix/libkiwix/issues/1099#issuecomment-2254454847) in the discussion of the issue.

**Pull request closed without merging**